pints/_optimisers/_nelder_mead.py

```diff
@@ -0,0 +1,337 @@
+#
+# Nelder-Mead simplex method.
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017-2019, University of Oxford.
+#  For licensing information, see the LICENSE file distributed with the PINTS
+#  software package.
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import numpy as np
+import pints
+
+
+# Flow
+#          Start
+#            |
+#         EVAL xr
+#            |       yes
+#          expand? ------> EVAL xe --> accept xe or xr
+#            |
+#            |no
+#            |       yes
+#         reflect? ------> accept xr
+#            |
+#            |no
+#            |
+#    contract in or out?
+#    |                 |
+#    |in               |out
+#    |                 |
+# EVAL xi           EVAL xo
+#    |                 |
+# better? ----+---- better?
+#    |        |       |
+#    |yes     |no      |yes
+#    |        |        |
+# accept xi   |     accept x
+#             |
+#          shrink
+#         EVAL n xs
+#
+# This becomes:
+#
+# Ask:
+#  if no xs set
+#    Set xs, ask for f(xs)
+#  elif shrink
+#    Ask for shrink points
+#  elif no xr set
+#    Order, set mean, set xr, ask for f(xr)
+#  else
+#    Ask for f(xp) (could be xi, xo, or xe)
+#
+# Tell:
+#  if no fs set
+#    Set fs, return (ask will set xr)
+#  if shrink:
+#    Shrink, unset xr, return (ask will set xr)
+#
+#  if no fr set
+#    set fr
+#  if expand
+#    if xp set
+#      accept xp or xr, unset xp&xr, return (ask will set xr)
+#    else
+#      set xp, return (ask will ask f(xp))
+#  elif contract
+#    if xp set
+#      if accept xp
+#        unset xp&xr, return (ask will set xr)
+#      else
+#        set shrink mode (ask will ask shrink points)
+#    else
+#      set xp, return (ask will ask f(xp))
+#  else
+#    accept xr, unset xr, return (ask will set xr)
+
+
+class NelderMead(pints.Optimiser):
+    """
+    Nelder-Mead downhill simplex method.
+
+    Implementation of the classical algorithm by [1], following the
+    presentation in Algorithm 8.1 of [2].
```

Comment:
Nicely documented! Do you think it's worth (later on; perhaps as a different optimiser) implementing Algorithm 8.2 since this corrects some of the pathologies of NM? In particular, it tries to guard against the simplex becoming of very small volume...

Reply:
It's probably a good idea yeah. Should we add a ticket? And do you think it should be a separate optimiser or an option or even just the default?

Reply:
Have made a ticket!